mantidproject/mantid
# Add JobTreeView widget

**Description of work**
Adds `JobTreeView` a spreadsheet like adaptation of QTreeView with an MVP style interface.
Part of the work for #22263.

**To test:**
1. Ensure passing builds
2. Apply the [attached patch](https://github.com/mantidproject/mantid/files/1995718/enable_bwi.txt) or modify the `mantid.properties.template` to enable the batch widget experimental interface.
3. Experiment with adding and removing different rows and navigating the table.
    **Keyboard Shortcuts**
    - `Ctrl-Enter` Create an empty new row as a child of the first row currently selected.
    - `Enter` Navigate to the row underneath the first row currently selected or else create one if there 
       is no such row at the current node in the tree.
    - `Shift-Enter` Navigate to the row above the first row currently selected or else do nothing if there 
       is no such row at the current node in the tree.
    - `Tab` Navigate the the cell on the right, or the first on the row below if there is no such cell. 
       Do nothing if there is no cell below or on the right at the current node in the tree.
    - `Shift-Tab` Navigate the the cell on the left, or the last on the row above if there is no such cell. 
       Do nothing if there is no cell above or on the left at the current node in the tree.
    - `Delete` Remove all rows currently selected and any of their children.
4. Type `Invalid` into a cell on the table and switch to a different cell. The cell containing `Invalid` 
    should have a red outline and an icon indicating the input is invalid. Ensure these effects are removed
    after changing it back again.
5. Filtering
    - Try typing a regular expression into the text box above the widget, this should filter the items based 
      on the text in their first column. Iff a node's first column text matches or if the node has a child whose first column matches then the node should be displayed.
    - While the row is filtered you should be able to edit the contents of the cells in the row. If editing the row causes it to no longer match the regex the row should continue to be displayed until adjusting the filter.
    - Attempting to insert new rows should not result in a crash, the filtering should be cleared and the row added as normal.
    - Attempting to delete a row should not result in a crash, the row should be deleted and the filter remain applied. Removing the root node from a filtered selection should not result in a crash unless it is the only node in the unfiltered tree.
6. Copy and Paste
    **Keyboard Shortcuts**
    - `Ctrl-C` Create an empty new row as a child of the first row currently selected.
    - `Ctrl-V` Navigate to the row above the first row currently selected or else do nothing if there 
       is no such row at the current node in the tree.
    Due to the lack of existing presenter/model implementations some cases of copy and paste aren't handled as you might expect. Ambiguous cases are also not handled at all, this results in a message like `Bad Selection For Copy` or `Bad Selection For Paste`. Test some of the currently handled situations with the scenarios below.

Create the tree shown below. 
```
A
   B
      C
D
```

Expand and select nodes A, B and C, press `Ctrl+C`. Select node D before pressing `Ctrl-V`. The tree 
should now look like ...

```
A
   B
      C
A
   B
      C
```

Create the tree shown below. 
```
A
   B
      C
D
E
```

Expand and select nodes A, B and C, press `Ctrl+C`. Select node D and E before pressing `Ctrl-V`. The tree should now look like the one below. Note the absence of the node E.

```
A
   B
      C
A
   B
      C
```

Finally create the tree shown below. 
```
A
   B
      C
D
   E
     F
G
```

Expand and select nodes A-F, press `Ctrl+C`. Select node G before pressing `Ctrl-V`. The tree should now look like. Note the absence of the node E.

```
A
   B
      C
D
   E
     F
A
   B
      C
D
   E
     F
```



Refs #22263

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
